jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java

```diff
@@ -0,0 +1,116 @@
+   * @param extraDirectory the source extra directory path
+   * @param extraDirectoryPermissions map from path on container to file permissions
+   * @return a {@link LayerConfiguration} for adding the extra directory to the container
+   * @throws IOException if walking the extra directory fails
+   */
+  public static LayerConfiguration extraDirectoryLayerConfiguration(
+      Path extraDirectory, Map<AbsoluteUnixPath, FilePermissions> extraDirectoryPermissions)
+      throws IOException {
+    LayerConfiguration.Builder builder =
+        LayerConfiguration.builder().setName(LayerType.EXTRA_FILES.getName());
+    new DirectoryWalker(extraDirectory)
+        .filterRoot()
+        .walk(
+            path -> {
+              AbsoluteUnixPath pathOnContainer =
+                  AbsoluteUnixPath.get("/").resolve(extraDirectory.relativize(path));
+              builder.addEntry(
+                  path, pathOnContainer, extraDirectoryPermissions.get(pathOnContainer));
+            });
+    return builder.build();
+  }
+
+  /**
+   * Constructs a new {@link JibContainerBuilder} for a WAR project.
+   *
+   * @param baseImage the base image of the container
+   * @param explodedWar the exploded WAR directory
+   * @param appRoot root directory in the image where the app will be placed
+   * @param extraFilesDirectory path to the source directory for the extra files layer
+   * @param extraDirectoryPermissions map from path on container to file permissions
+   * @return {@link JibContainerBuilder} containing the layers for the exploded WAR
+   * @throws IOException if adding layer contents fails
+   */
+  public static JibContainerBuilder fromExplodedWar(
+      RegistryImage baseImage,
+      Path explodedWar,
+      AbsoluteUnixPath appRoot,
+      Path extraFilesDirectory,
+      Map<AbsoluteUnixPath, FilePermissions> extraDirectoryPermissions)
+      throws IOException {
+    Path webInfLib = explodedWar.resolve("WEB-INF/lib");
+    Path webInfClasses = explodedWar.resolve("WEB-INF/classes");
+    Predicate<Path> isDependency = path -> path.startsWith(webInfLib);
+    Predicate<Path> isClassFile =
+        path -> path.startsWith(webInfClasses) && path.getFileName().toString().endsWith(".class");
+    Predicate<Path> isResource = isDependency.or(isClassFile).negate();
+
+    JavaContainerBuilder javaContainerBuilder =
+        JavaContainerBuilder.from(baseImage)
+            .setAppRoot(appRoot)
+            .setResourcesDestination(RelativeUnixPath.get(""))
+            .setClassesDestination(RelativeUnixPath.get("WEB-INF/classes"))
+            .setDependenciesDestination(RelativeUnixPath.get("WEB-INF/lib"));
+
+    if (Files.exists(explodedWar)) {
+      javaContainerBuilder.addResources(explodedWar, isResource);
+    }
+    if (Files.exists(webInfClasses)) {
+      javaContainerBuilder.addClasses(webInfClasses, isClassFile);
+    }
+    if (Files.exists(webInfLib)) {
+      javaContainerBuilder.addDependencies(webInfLib, isDependency);
```

Comment:
`isDependency`, which checks if the given path starts with `WEB-INF/lib`, seems unnecessary to pass, since the directory root we pass here is `webInfLib`. Then, I guess you don't need to introduce the new `addDependencies()` method?

Reply:
Oh, I see you still need to pass a list of JARs. But the filter is not necessary.

Reply:
The filter is used earlier to define the resources filter (https://github.com/GoogleContainerTools/jib/pull/1572/files/5c0b824d12984cefefa3f2dd90b567a9869f95e5#diff-945af34eb6dbffe51159846ba6bfa9ccR85), but yeah I guess it's not necessary here.

I think having an `addDependencies` method that takes a filter is still nice for consistency among the other methods. Maybe I should even add another method that takes no filter and walks a directory, much like the classes/resources methods?

Reply:
This approach is mildly dangerous in that the user might provide a set of filters that misses some files, such that some dependencies don't make it through.  We're trying to intelligently place the file contents in appropriate layers, rather than filter out files.  As it's hard to prove our filters don't miss anything, I wonder if it might not be safer to walk the files with a function that returns a layer-indicator for each file.

Reply:
@briandealwis I remember a similar, if not same, idea: https://github.com/GoogleContainerTools/jib/issues/1065

From what I remember, I tried the idea, but there were some problems which I don't remember exactly. Things might have changed though.